liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseMojo.java

```diff
@@ -0,0 +1,163 @@
+                // then do not overwrite
+                //
+                if (! projectProperties.containsKey(workKey)) {
+                    projectProperties.put(workKey, entry.getValue());
+                }
+                this.configuration.put(key, entry.getValue());
+            }
+        }
+    }
+
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        Main.runningFromNewCli = true;
+
+        String goal = mojoExecution.getGoal();
+        LiquibaseConfiguration liquibaseConfiguration = Scope.getCurrentScope().getSingleton(LiquibaseConfiguration.class);
+        MojoConfigurationValueProvider mojoConfigurationValueProvider = null;
+        try {
+            mojoConfigurationValueProvider = new MojoConfigurationValueProvider(configuration);
+            liquibaseConfiguration.registerProvider(mojoConfigurationValueProvider);
+
+            Map<String, Object> scopeValues = new HashMap<>();
+            scopeValues.put(Scope.Attr.resourceAccessor.name(), new MavenResourceAccessor(project));
+            CommandScope commandScope = new CommandScope(goal);
+            Scope.child(scopeValues, () -> {
+                commandScope.execute();
+            });
+        } catch (Exception e) {
+            throw new MojoExecutionException(e.getMessage(), e);
+        } finally {
+            if (mojoConfigurationValueProvider != null) {
+                liquibaseConfiguration.unregisterProvider(mojoConfigurationValueProvider);
+                mojoConfigurationValueProvider = null;
+            }
+        }
+    }
+
+    //
+    // This class provides values which are on the command line or in the POM
+    //
+    private static class MojoConfigurationValueProvider extends AbstractMapConfigurationValueProvider {
+        private final Map<String, Object> configurationMap;
+
+        MojoConfigurationValueProvider(Map<String, Object> configurationMap) {
+            this.configurationMap = configurationMap;
+        }
+
+        @Override
+        protected Map<?, ?> getMap() {
+            return configurationMap;
+        }
+
+        /**
+         *
+         * This override of keyMatches allows the global arguments
+         * to be specified as either "liquibase.arg" or "arg"
+         *
+         * @param   wantedKey the configuration key requested
+         * @param   storedKey the key stored in the map
+         * @return  boolean
+         *
+         */
+        @Override
+        protected boolean keyMatches(String wantedKey, String storedKey) {
+            if (super.keyMatches(wantedKey, storedKey)) {
+                return true;
+            }
+            if (wantedKey.startsWith("liquibase.command.")) {
+                return super.keyMatches(wantedKey.replaceFirst("^liquibase\\.command\\.", ""), storedKey);
+            }
+
+            return super.keyMatches(wantedKey.replaceFirst("^liquibase\\.", ""), storedKey);
```

Comment:
I know I've personally duplicated this logic in another place already. Do you think it would be possible to move this to an abstract class and remove the duplication here and elsewhere?

Reply:
I had started on #3286 which was part of my same "how can maven work with the command framework" spike as this, but I kept separate because it was "how do we add an 'integration api' layer" initial experimentation. A big goal of that is to reduce duplicated logic like this. 

It's probably worth keeping the duplication for now and getting a command-API version of maven out there. Then use the 3rd integration we convert (ant?) as the time we find and extract the shared logic from all three integrations.